core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -987,6 +987,25 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
         return Optional.empty();
     }
 
+    /**
+     * Attempt to push down the join operation.
+     * <p>
+     * Connectors can indicate whether they don't support join pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}. Connectors should expect this method may be called multiple times.
+     * </p>
+     */
+    default Optional<ConnectorTableHandle> applyJoin(
```

Comment:
I am not sure this return type is sufficient, but i didn't figure out the cases yet.